### PR TITLE
ARROW-9874: [C++] Add sink-owning version of IPC writers

### PR DESCRIPTION
**Pull request by pitrou**
Also rename `New(File|Stream)Writer` to `Make(File|Stream)Writer`

### CODE REVIEW

File: cpp/examples/minimal_build/example.cc [hunk 1]
```diff
@@ -52,8 +52,8 @@ Status RunMain(int argc, char** argv) {
   ARROW_ASSIGN_OR_RAISE(auto output_file,
                         arrow::io::FileOutputStream::Open(arrow_filename));
   ARROW_ASSIGN_OR_RAISE(auto batch_writer,
-                        arrow::ipc::NewFileWriter(output_file.get(),
-                                                  table->schema()));
+                        arrow::ipc::MakeFileWriter(output_file.get(),
```
Comment (kou) [inline cpp/examples/minimal_build/example.cc:55]:
```suggestion
                        arrow::ipc::MakeFileWriter(output_file,
```

File: cpp/src/arrow/ipc/read_write_test.cc [hunk 2]
```diff
@@ -903,7 +903,8 @@ struct FileWriterHelper {
 
     ARROW_ASSIGN_OR_RAISE(buffer_, AllocateResizableBuffer(0));
     sink_.reset(new io::BufferOutputStream(buffer_));
-    ARROW_ASSIGN_OR_RAISE(writer_, NewFileWriter(sink_.get(), schema, options, metadata));
+    ARROW_ASSIGN_OR_RAISE(writer_,
+                          MakeFileWriter(sink_.get(), schema, options, metadata));
```
Comment (kou) [inline cpp/src/arrow/ipc/read_write_test.cc:907]:
```suggestion
                          MakeFileWriter(sink_, schema, options, metadata));
```

File: cpp/src/arrow/ipc/read_write_test.cc [hunk 3]
```diff
@@ -968,7 +969,7 @@ struct StreamWriterHelper {
   Status Init(const std::shared_ptr<Schema>& schema, const IpcWriteOptions& options) {
     ARROW_ASSIGN_OR_RAISE(buffer_, AllocateResizableBuffer(0));
     sink_.reset(new io::BufferOutputStream(buffer_));
-    ARROW_ASSIGN_OR_RAISE(writer_, NewStreamWriter(sink_.get(), schema, options));
+    ARROW_ASSIGN_OR_RAISE(writer_, MakeStreamWriter(sink_.get(), schema, options));
```
Comment (kou) [inline cpp/src/arrow/ipc/read_write_test.cc:972]:
```suggestion
    ARROW_ASSIGN_OR_RAISE(writer_, MakeStreamWriter(sink_, schema, options));
```

File: cpp/src/arrow/ipc/writer.cc [hunk 4]
```diff
@@ -1042,8 +1042,13 @@ class ARROW_EXPORT IpcFormatWriter : public RecordBatchWriter {
 
 class StreamBookKeeper {
  public:
-  explicit StreamBookKeeper(const IpcWriteOptions& options, io::OutputStream* sink)
+  StreamBookKeeper(const IpcWriteOptions& options, io::OutputStream* sink)
       : options_(options), sink_(sink), position_(-1) {}
```
Comment (kou) [inline cpp/src/arrow/ipc/writer.cc:1046]: Do we need `owned_sink_(nullptr)` here?

**Review by kou — APPROVED** [review]
+1

File: cpp/src/arrow/ipc/writer.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/arrow/ipc/writer.cc:1046]: No, it's implicit (a `shared_ptr` is always initialized).

File: cpp/src/arrow/ipc/read_write_test.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/ipc/read_write_test.cc:907]: `sink_` is a `unique_ptr` here, so not possible.

File: cpp/src/arrow/ipc/read_write_test.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/ipc/read_write_test.cc:972]: Same here.

File: cpp/examples/minimal_build/example.cc (hunk 1 above)
Comment (pitrou) [reply cpp/examples/minimal_build/example.cc:55]: Will do, thanks.

**Pull request closed without merging**